dvc/objects/stage.py

```diff
@@ -49,43 +66,64 @@ def get_file_hash(path_info, fs, name, state=None):
     return hash_info
 
 
-def _calculate_hashes(path_info, fs, name, state, **kwargs):
-    def _hash(file_info):
-        return file_info, get_file_hash(file_info, fs, name, state)
+def _get_file_obj(path_info, fs, name, odb=None, state=None, upload=False):
+    if upload:
+        assert odb and name == "md5"
+        return _upload_file(path_info, fs, odb)
+
+    obj = HashFile(
+        path_info, fs, get_file_hash(path_info, fs, name, state=state)
+    )
+    return path_info, obj
+
 
+def _build_objects(path_info, fs, name, odb, state, upload, **kwargs):
     with Tqdm(
-        unit="md5", desc="Computing file/dir hashes (only done once)",
+        unit="md5",
+        desc="Computing file/dir hashes (only done once)",
+        disable=kwargs.pop("no_progress_bar", False),
     ) as pbar:
-        worker = pbar.wrap_fn(_hash)
-        with ThreadPoolExecutor(max_workers=fs.hash_jobs) as executor:
-            pairs = executor.map(worker, fs.walk_files(path_info, **kwargs))
-            return dict(pairs)
+        worker = pbar.wrap_fn(
+            partial(
+                _get_file_obj,
+                fs=fs,
+                name=name,
+                odb=odb,
+                state=state,
+                upload=upload,
+            )
+        )
+        with ThreadPoolExecutor(
+            max_workers=kwargs.pop("jobs", fs.hash_jobs)
```

Comment:
Btw, looks like we've lost
```
jobs = jobs or min((from_fs.jobs, odb.fs.jobs))
```
we used to have for transfer. We could bring it back here with `if upload` for now, to match previous behavior. But I wonder of there is a reasonable common approach here in general, looks like this finally pushes us to reconsider it (again, just a thought).

Reply:
> we used to have for transfer. We could bring it back here with if upload for now, to match previous behavior. But I wonder of there is a reasonable common approach here in general, looks like this finally pushes us to reconsider it (again, just a thought).

I also thought about it, and I think it makes sense to use `fs.hash_jobs` since we are actively calculating the hash in the meanwhile (but still the IO time is 2X than the normal [if you include upload], so maybe on `uplaod` it can be `fs.hash_jobs * 2`)

Reply:
If we think about it, both jobs and hash_jobs are usually just N*CPU heuristics. Both uploading and hashing are likely IO bound, so there is probably a common number of jobs that benefits them. Once again this is probably a topic for research for after fsspec migration, since it will make it easier to experiment with that. But it is good that we are keeping it in mind.